src/chainstate/stacks/boot/exit-at-rc.clar

```diff
@@ -0,0 +1,199 @@
+
+;; This function can only be called once, when it boots up
+(define-public (set-burnchain-parameters (first-burn-height uint) (reward-cycle-length uint) (min-exit-rc uint))
+    (begin
+        (asserts! (not (var-get configured)) (err ERR_NOT_ALLOWED))
+        (var-set first-burnchain-block-height first-burn-height)
+        (var-set pox-reward-cycle-length reward-cycle-length)
+        (var-set absolute-minimum-exit-rc min-exit-rc)
+        (var-set configured true)
+        (ok true))
+)
+
+;; Stores number of votes for exit proposals by coupling it with the reward cycle the vote is cast in.
+(define-map rc-proposal-votes
+    {
+        proposed-rc: uint,
+        curr-rc: uint
+    }
+    { votes: uint }
+)
+
+;; Stores number of vetos for exit proposals by coupling it with the reward cycle the veto is cast in.
+(define-map rc-proposal-vetoes
+    {
+        proposed-rc: uint,
+        curr-rc: uint
+    }
+    { vetoes: uint }
+)
+
+;; Keeps track of miner vetos; used to ensure that a miner cannot cast multiple vetos in the same block.
+(define-map exercised-veto
+    {
+        proposed-rc: uint,
+        veto-height: uint
+    }
+    { vetoed: bool }
+)
+
+;; Keeps track of voter specific information; used to ensure a voter doesn't vote twice in the same stacking period.
+(define-map voter-state
+    { address: principal }
+    {
+        proposed-rc: uint,
+        amount: uint,
+        expiration-reward-cycle: uint
+    }
+)
+
+;; What's the reward cycle number of the burnchain block height?
+;; Will runtime-abort if height is less than the first burnchain block (this is intentional).
+;; Returns uint
+(define-private (burn-height-to-reward-cycle (height uint))
+    (let
+        (
+            (local-first-burnchain-block-height (var-get first-burnchain-block-height))
+            (local-pox-reward-cycle-length (var-get pox-reward-cycle-length))
+        )
+
+        (asserts! (> height local-first-burnchain-block-height) (err ERR_BURN_BLOCK_HEIGHT_TOO_LOW))
+        (asserts! (> local-pox-reward-cycle-length u0) (err ERR_AMOUNT_NOT_POSITIVE))
+
+        (ok (/ (- height local-first-burnchain-block-height) local-pox-reward-cycle-length))
+    )
+)
+
+;; What's the current PoX reward cycle?
+;; Returns uint
+(define-private (current-pox-reward-cycle)
+    (burn-height-to-reward-cycle burn-block-height))
+
+
+;; For a specific reward cycle, this function tried to add "amount" number of votes for the proposed exit reward cycle.
+;; Note: votes are tied to reward cycles, since the vote's lifetime is tied to the voter's stacking duration.
+(define-private (add-to-rc-proposal-map (cycle-opt (optional uint)) (proposed-rc uint) (amount uint))
+    (begin
+        (match cycle-opt
+            cycle (match (map-get? rc-proposal-votes {proposed-rc: proposed-rc, curr-rc: cycle})
+                              existing-votes (map-set rc-proposal-votes {proposed-rc: proposed-rc, curr-rc: cycle} {votes: (+ amount (get votes existing-votes)) })
```

Comment:
Minor style nit: you can make this more succinct with `merge`:

```
existing-votes (map-set rc-proposal-votes (merge existing-votes { votes: (+ amount (get votes existing-votes)) }))
```

Reply:
It looks like `merge` just returns the value tuple. I would need to re-write the key tuple regardless, so I don't think it saves space here.